Prototype API for fetching os cacerts

dgud opened the pull request:
API: load/0 load/1 and get/0

Writes to persistent_term to cache the results,
user can re-load cache with load/0 and load/1 functions.

load/1 is made for users with an unsupported OS, or if they
want to load the cache with data from another file.

Currently works on linux (ubuntu) and win32

u3s reviewed (commented):
i've read the code and have no comments.

max-au reviewed (commented):
I was thinking of any way to test this, and found nothing else except listing a bunch of actual Root CA thumbprints/names/... in the test case. Given that Root CA changes are extremely rare it might be okay to see a test case failing whenever that happens.

dgud commented on lib/public_key/src/pubkey_os_cacerts.erl:
```diff
@@ -0,0 +1,184 @@
+load_darwin() ->
+    %% Temporary solution, should probably be re-written to use Keychain Access API
+    KeyChainFile = "/System/Library/Keychains/SystemRootCertificates.keychain",
+    TmpPemFile = "/tmp/all_certs.pem",
```
Probably not, this should be good enough, tm.
Please do, this will remain until someone with Obj-C knowledge decides that it is a much better solution,
and updates the nif, writes a configure file and ..

lukebakken replied:
Does the `security export` command's `-o` switch allow `-` as an argument to export to stdout, removing the use of a temp file? I'd check but I don't have access to OS X.

wojtekmach replied:
Going through Security API would be ideal but in the meantime I think we can improve how we shell out. The default is to write to stdout so going through the temp file should be unnecessary:

```erl
> X = os:cmd("security export -t certs -f pemseq -k /System/Library/Keychains/SystemRootCertificates.keychain"), {lists:sublist(X,10), length(X)}.
{"-----BEGIN",249291}
```

dgud commented on lib/public_key/src/pubkey_os_cacerts.erl:
```diff
@@ -0,0 +1,184 @@
+%% Return cacerts
+-spec get() -> [public_key:combined_cert()].
+get() ->
+    case persistent_term:get(?MODULE, not_loaded) of
```
> If I understand correctly, this persistent_term would stay in memory forever

True, but do we really need to handle that case? 
I.e. add a delete_cacerts() function to the API that will never be called, or do you see another solution.

There is no automated reload, which is what I wanted:

1.   We don't want change the public_key from a library application

2.  I think I implemented the original cache in ssl and Ingela have been fixing bugs in that since then.
     So I want to keep it simple.
    The user (should) know when the system is updated and can thus call **load/[0|1]**

tsloughter commented:
This is great to see being added.

My only thought is not really a concern for this specific PR but I think it'd be nice if `httpc` had an option (or even be the default) to use these instead of having to do what we have had to do before of passing a long list of `ssl` args including a full list of cacerts. Security WG wrote this up https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/inets

In rebar3 we construct the ssl options for `httpc` here https://github.com/erlang/rebar3/blob/main/src/rebar_utils.erl#L1080-L1081

We'd still need to override in case the user specifies a specific cacertfile in the configuration, but for general usage it'd be nice to not have to configure this long list of options to make a secure HTTPS request.

dgud commented:
@max-au   Extended the tests after your suggestions and added a cacerts_clear() for completion,
I also changed cacerts_load/1 to take one file as argument instead of a list of tries.

@tsloughter  Yeah that was my intention but should be done but outside of this PR.
You could use public_key:cacerts_load(File) before calling httpc and that would automatically use those certs.

emixa-d commented:
Not really an Erlang user myself, but I stumbled upon this patch from
[https://issues.guix.gnu.org/54796#52](https://issues.guix.gnu.org/54796#52)
Could the `SSL_CERT_FILE` environment variable be supported?
That way, non-privileged users could easily override which certificates
to trust.

wojtekmach commented:
> Could the `SSL_CERT_FILE` environment variable be supported? That way, non-privileged users could easily override which certificates to trust.

Perhaps worth mentioning that curl supports [`SSL_CERT_FILE`](https://curl.se/docs/manpage.html#SSLCERTFILE) and [`CURL_CA_BUNDLE`](https://curl.se/docs/manpage.html#--cacert) environment variables too.

max-au commented:
Thanks @dgud !
Looking more and more at this code, I feel more and more convinced that it belongs `ssl` application.
1. There is already a NIF (the one that used openssl crypto). It will save `public_key` from adding NIF code.
2. The only way to use Root CA list is actually in "ssl" application.
3. There is already a PEM cache in SSL application - and it already has (battle-tested) eviction logic.

I also suspect that ssl_pem_cache is where all the certificates end up anyway (unless cache is disabled, but in this case performance degrades too much).

One approach I can think of is adding a `ssl_root_ca_store` to `ssl_admin_sup`. This would allow to have this store addressable by the certificate thumbnail or any other mechanism that `ssl` app needs to find the certificate chain.

lukebakken commented:
Supporting environment variables should be out of the scope of providing easy access to the system's trusted root certificates.

emixa-d commented:
Luke Bakken schreef op zo 10-04-2022 om 12:29 [-0700]:
> Supporting environment variables should be out of the scope of
> providing easy access to the system's trusted root certificates.

On my system, /etc/ssl/certs sometimes does not exist:

```
# imagine openssl was replaced by some erlang thing here
$ guix shell --container coreutils nss-certs openssl
[env]$ ls /etc
group  hosts  passwd
[env]$ ls /etc/ssl
ls: cannot access '/etc/ssl': No such file or directory
```

instead, to find the root certificates, you have to look in
SSL_CERT_DIR or SSL_CERT_FILE:

```
[env]$ echo $SSL_CERT_DIR
/gnu/store/3hbqsd3h44fl9ri8wsf4nn56dhh15c51-profile/etc/ssl/certs
[env]$ ls $SSL_CERT_DIR
[lots of certificates]
```

so hardcoding /etc/ssl/certs is not always sufficient.

lukebakken commented:
@emixa-d thanks for pointing that out. It seems like it would be useful to investigate the following:
* How does a common utility like `curl` determine CA file / directory locations?
* Are `SSL_CERT_FILE` and `SSL_CERT_DIR` "standard" names?
* What do other VMs do? Java, for instance.
* What are the standard locations for a few major Linux distros and BSDs?

emixa-d commented:
Dan Gudmundsson schreef op di 12-04-2022 om 00:33 [-0700]:
> > Not really an Erlang user myself, but I stumbled upon this patch
> > from
> > https://issues.guix.gnu.org/54796#52
> > Could the SSL_CERT_FILE environment variable be supported?
> > That way, non-privileged users could easily override which
> > certificates
> > to trust.
> I do like that from a useability perspective, but we have had
> complaints before for that an erlang system
> default loads an '.erlang' file and other things like that and that
> that can be an security issue that the user
> it allowed to change the behavior by default.

I don't follow how this can be a security issue (assuming this is not
about setuid or setgid binaries or the like) -- if a user is malicious,
they could just run a patched erlang that recognises SSL_CERT_DIR. 
E.g., not sure how the erlang VM is actually started, but it on Guix it
would be something along the lines of:

  $ SSL_CERT_FILE=$HOME/evil.crt guix shell erlang --with-source=$HOME/my-patches-erlang.tar.gz -- erlang evil.erlang

This can also be done with an unpatched erlang: just run erlang in a
a container that bind-mounts /etc/ssl/certs to $HOME/evil-certs.

The hypothetical malicious user could also modify which certificates are
used, without using a patched erlang or containers:

  1. grab a copy of the source code of the erlang project
  2. modify it to look in $HOME/evil-certs
  3. compile & run

They could also do their maliciousity without Erlang, e.g. with C and gcc.

emixa-d commented:
(GitHub seems to have lost this response)

Luke Bakken schreef op zo 10-04-2022 om 15:48 [-0700]:
> @emixa-d thanks for pointing that out. It seems like it would be useful
> to investigate the following:
> * How does a common utility like curl determine CA file / directory
> locations?

The _utility_ curl looks in `CURL_CA_BUNDLE` (see
https://curl.se/docs/sslcerts.html).  I'd assume it fallbacks
to /etc/ssl/certs but I don't actually know that for a fact.

AFAIK, the _library_ curl does not look in any environemn variables by
default.  However, in Guix (see gnu/packages/patches/curl-use-ssl-cert-
env.patch), libcurl is patched to look in `SSL_CERT_DIR`/`SSL_CERT_FILE` by
default.

 * Are `SSL_CERT_FILE` and `SSL_CERT_DIR` "standard" names?

`SSL_CERT_DIR` / `SSL_CERT_FILE` are used by OpenSSL.
GNUtls currently does not support any environment variables.
However, there is a patch adding support for `SSL_CERT_DIR` /
`SSL_CERT_FILE`:
<https://gitlab.com/gnutls/gnutls/-/merge_requests/1541>.

Guix (as in, some of the support libraries like guix/git.scm and
guix/build/download.scm, not the set of packages) looks for
certificates in `SSL_CERT_FILE`/`SSL_CERT_DIR`, with a fallback to
/etc/ssl/certs/ca-certificates.crt and /etc/ssl/certs/ca-bundle.crt.

I am not aware of any ‘competing’ standards.

Git has its own environment variable `GIT_SSL_CAINFO`.

'nss' has its own mechanism  (some database and not directly files --
probably not relevant to
erlang/otp?)(https://www.happyassassin.net/posts/2015/01/12/a-note-about-ssltls-trusted-certificate-stores-and-platforms/).

 * What do other VMs do? Java, for instance.

If Java recognises any environment variables, I haven't found them.
According to
<https://stackoverflow.com/questions/5871279/ssl-and-cert-keystore#>,
it looks in $JAVA_HOME/lib/security/jssecacerts or
$JAVA_HOME/lib/security/cacerts which can be overriden by setting
‘javax.net.ssl.trustStore.’

Firefox does not use `SSL_CERT_FILE`/`SSL_CERT_DIR`.  It has its own
certificate database, named 'nss'.

I'm out of VMs, unless QEMU counts here.

* What are the standard locations for a few major Linux distros and
BSDs?

On Guix System, it's /etc/ssl/certs, but only if not running in some
container made with "guix shell --container, and only if the system
administrator installed a certificate globally (or, in Guix
terminology, installed a certificate package in the system profile).
Guix System also sets SSL_CERT_DIR/SSL_CERT_FILE.

On Debian systems, it's /etc/ssl/certs.

Apparently, on at least one ArchLinux version, it is /etc/ssl/cert.pem
(https://github.com/NixOS/nixpkgs/issues/8247#issuecomment-179364714)
They also mention /etc/ssl/certs/ca-certificates.crt,
/etc/ssl/certs/ca-bundle.crt, /etc/pki/tls/certs/ca-bundle.crt.
Apparently these files are created by NixOS?

OpenSUSE puts it in yet another location: /var/lib/ca-
certificates/bundle.pem
(https://www.happyassassin.net/posts/2015/01/12/a-note-about-ssltls-trusted-certificate-stores-and-platforms/).
 There's also another location but it's broken or something?  That blog
post recommends just using SSL_CTX_set_default_verify_paths() (assuming
you are using openssl)

Apparently FreeBSD puts things in /etc/ssl/cert.pem:
<https://www.freshports.org/security/ca_root_nss>.

dgud commented:
@emixa-d  I believe the discussion about security comes from that if you download a binary and run it
it should do what you expect, e.g. in this case only fetch data from servers that can be verified by OS cacerts,
not the certs the user set or a hacker on the user system.

lukebakken commented:
> The load/1 function can be used for loading an arbitrary file, so the user/application can easily read the environment
variable and load the file.

@dgud thanks. I agree, environment variable support should be left up to application developers.

dgud commented:
@max-au @lukebakken @wojtekmach  Removed the temporary file and read from stdin instead.

wojtekmach commented:
> Removed the temporary file and read from stdin instead.

Works great!

potatosalad commented:
> I feel more and more convinced that it belongs `ssl` application.

@max-au As a counter-argument: I think the scope of the `public_key` library is all-things-PKI-related, which is where I would expect to find things like the OS Root Certs.

@dgud Out of curiosity, why add a new NIF to `public_key` versus adding a new NIF function to the existing `crypto` NIF while keeping the nice-to-use API in `public_key`?  That seems to be the pattern for things like `crypto:sign/4,5` and `public_key:sign/3,4`.

dgud commented:
@potatosalad Because crypto is "thin" wrapper towards "SSLs crypto-lib" and unrelated to this.
It would have been simpler but kind of a hack, IMO.

emixa-d commented:
(I previously sent this response GitHub by e-mail but it seems to have lost this response)

Dan Gudmundsson schreef op di 12-04-2022 om 13:34 [+0000]:
> @emixa-d I believe the discussion about security comes from that if
> you download a binary and run it
> it should do what you expect, e.g. in this case only fetch data from
> servers that can be verified by OS cacerts,
> not the certs the user set or a hacker on the user system.

If I, as user, download some binaries and run them, I expect them to
use the cacerts I specified as trusted via the environment variables,
instead of the OS' ~128 fallback root CAs (in case of Mozilla's set,
about 100 points of failure!) I don't actually trust but have to use to
actually browse the web.

If some attacker has compromised my shell process to set some
environment variables then I have bigger problems to worry about.
Also, the attacker could just set `LD_LIBRARY_PATH` or `LD_PRELOAD` or just directly run some malicious binary, why would they bother with `SSL_CERT_FILE` and the like?

Anyway, OK, but I expect it to be eventually patched downstream because
...

> The load/1 function can be used for loading an arbitrary file, so the
> user/application can easily read the environment
> variable and load the file.
> 
> @dgud thanks. I agree, environment variable support should be left up
> to application developers.
> 
> 

... AFAICT, often applications don't bother with opting-into
environment variables if by default things use the global certificate
store (/etc/ssl/certs or the like).  Or maybe I'll be pleasantly
surprised.
